### PR TITLE
Support Binary IO for Arrays

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
towards https://github.com/lfortran/lfortran/issues/2001

### CODE REVIEW

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 1]
```diff
@@ -1808,6 +1809,34 @@ LFORTRAN_API void _lfortran_read_int32(int32_t *p, int32_t unit_num)
     }
 }
 
+LFORTRAN_API void _lfortran_read_array_int32(int32_t *p, int32_t unit_num)
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c:1812]: I think you need to pass in how much to read, don't you?

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 2]
```diff
@@ -1808,6 +1809,34 @@ LFORTRAN_API void _lfortran_read_int32(int32_t *p, int32_t unit_num)
...
+    }
+
+    bool unit_file_bin;
+    FILE* filep = get_file_pointer_from_unit(unit_num, &unit_file_bin);
+    if (!filep) {
+        printf("No file found with given unit\n");
+        exit(1);
+    }
+
+    int n = sizeof(p);
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c:1830]: Right now it is reading 8 bytes (`sizeof(p)` which is always equal to 8). We need to be reading "n" bytes, the length of the array.

Comment (Shaikh-Ubaid) [reply src/libasr/runtime/lfortran_intrinsics.c:1830]: I think you are right. I thought `sizeof()` for arrays returns the total size of the array. I just checked here https://onlinegdb.com/0TLS5hw4l and it seems it only happens for statically allocated arrays.

**Comment by certik** [comment]
I tested 057ee84ec46b5c2b17c6e17d546080ece56f24ca with fastGPT and I think it will work. There is a bug for having the array inside a derived type, but a workaround of using a local array works! Once this is merged, I'll figure out how to get the derived type working.

**Comment by Shaikh-Ubaid** [comment]
>I tested https://github.com/lfortran/lfortran/commit/057ee84ec46b5c2b17c6e17d546080ece56f24ca with fastGPT and I think it will work. There is a bug for having the array inside a derived type, but a workaround of using a local array works! Once this is merged, I'll figure out how to get the derived type working.

This PR adds binary io support for arrays of integers and reals. It seems fastGPT also uses arrays of characters. I think we need to support it as well.

**Comment by certik** [comment]
Yes. I only tested the real so far. Yes, we need characters too.

**Comment by Shaikh-Ubaid** [comment]
I think specifically we need to support binary io for these two character arrays https://github.com/certik/fastGPT/blob/1ddf333a93171947cec142583aa30b9273413c3d/gpt2.f90#L24.

File: src/libasr/runtime/lfortran_intrinsics.c (hunk 2 above)
Comment (Shaikh-Ubaid) [reply src/libasr/runtime/lfortran_intrinsics.c]: Thank you so much for this. This was very helpful.

**Comment by certik** [comment]
Yes, correct. Do you know how to do it?

**Comment by Shaikh-Ubaid** [comment]
It seems there is squash option being shown here on GitHub UI. I have rebased this PR. Thus, I think we can merge this PR normally.

**Comment by Shaikh-Ubaid** [comment]
>I think specifically we need to support binary io for these two character arrays https://github.com/certik/fastGPT/blob/1ddf333a93171947cec142583aa30b9273413c3d/gpt2.f90#L24.

I was hoping to support it in this PR (for example the commit [Support IO for allocatable strings](https://github.com/lfortran/lfortran/pull/2036/commits/fbdd8b037f8edd7ce95fd4554606af81a160d589) is in that direction). It seemed to take longer so I dropped it for this PR and supported only integers and reals.

>Yes, correct. Do you know how to do it?

I need to try and see. It might take some attempts. If we have less time, you can take it up. I can focus on other issues at https://github.com/lfortran/lfortran/issues/1997.

**Comment by Shaikh-Ubaid** [comment]
@certik This is ready. Also, please see https://github.com/lfortran/lfortran/pull/2036#issuecomment-1646718763.

File: tests/reference/llvm-string_04-2d3483c.stdout [hunk 3]
```diff
@@ -354,7 +354,7 @@ loop.body:                                        ; preds = %loop.head
 loop.end:                                         ; preds = %loop.head
   call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([3 x i8], [3 x i8]* @17, i32 0, i32 0), i8* getelementptr inbounds ([2 x i8], [2 x i8]* @16, i32 0, i32 0))
   call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([3 x i8], [3 x i8]* @20, i32 0, i32 0), i8* getelementptr inbounds ([2 x i8], [2 x i8]* @19, i32 0, i32 0))
-  call void (i8**, i32, ...) @_lfortran_alloc(i8** %str, i32 64)
+  call void (i8**, i32, ...) @_lfortran_alloc(i8** %str, i32 8)
```
Comment (certik) [inline tests/reference/llvm-string_04-2d3483c.stdout:357]: I checked and this is a correct change. It's coming from `allocate(character(len=8)::str)`, so the length should be 8, not 64.

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 4]
```diff
@@ -1380,6 +1380,7 @@ LFORTRAN_API void _lfortran_free(char* ptr) {
 
 LFORTRAN_API void _lfortran_alloc(char** ptr, int32_t size) {
     *ptr = (char *) malloc(size);
+    memset(*ptr, ' ', size);
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c:1383]: Is this needed? I think we should not zero the memory.

File: tests/reference/llvm-string_04-2d3483c.stdout (hunk 3 above)
Comment (Shaikh-Ubaid) [reply tests/reference/llvm-string_04-2d3483c.stdout:357]: Yes.

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 5]
```diff
@@ -1824,9 +1825,10 @@ LFORTRAN_API void _lfortran_read_char(char **p, int32_t unit_num)
         exit(1);
     }
 
-    *p = (char*)malloc(strlen(*p) * sizeof(char));
+    int n = strlen(*p);
+    *p = (char*)malloc(n * sizeof(char));
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:1855]: .

File: src/libasr/runtime/lfortran_intrinsics.c (hunk 4 above)
Comment (Shaikh-Ubaid) [reply src/libasr/runtime/lfortran_intrinsics.c:1383]: It was being for this https://github.com/lfortran/lfortran/pull/2036/commits/fbdd8b037f8edd7ce95fd4554606af81a160d589#r1271368142. `strlen(*p)` returns zero for unallocated memory.

Comment (certik) [reply src/libasr/runtime/lfortran_intrinsics.c]: Tests pass for me, so I removed it in 3a966318dcac2d7043058992eba38a8da4533e87.

Comment (Shaikh-Ubaid) [reply src/libasr/runtime/lfortran_intrinsics.c]: If it is fine with you, shall we tackle it in a separate PR? Maybe we can open an issue it.

**Review by certik — APPROVED** [review]
I think this looks good. Let's merge it and I'll start working with this.

Let's implement the character arrays in the next PR.

**Comment by Shaikh-Ubaid** [comment]
@certik Is it squash merge for this PR or normal merge?

**Comment by certik** [comment]
Normal merge.

**Comment by certik** [comment]
Actually, the failure of a struct I hit above is unrelated. I think this works completely for fastGPT.

Just the character must be done I think.

**Pull request merged**